reposerver/repository/repository.go

```diff
@@ -2843,6 +2843,19 @@ func (s *Service) UpdateRevisionForPaths(_ context.Context, request *apiclient.U
 	s.metricsServer.IncPendingRepoRequest(repo.Repo)
 	defer s.metricsServer.DecPendingRepoRequest(repo.Repo)
 
+	// Check if the synced revision is present in the repo. If not, checkout the revision
+	// Note that fetching with an explicit version can cause repo bloat [https://github.com/argoproj/argo-cd/issues/8845],
+	// so we should take lock (theoretically not necessary) and checkout.
+	if !gitClient.IsRevisionPresent(syncedRevision) {
+		closer, err := s.repoLock.Lock(gitClient.Root(), syncedRevision, true, func() (goio.Closer, error) {
```

Comment:
Taking this lock should be considered carefully because of the performance aspect, but this happens in rare situations.